File: zZ/java/palindrome-partitioning.java
/**
 * Palindrome Partitioning
 * 
 * Given a string s, partition s such that every substring of the partition is a palindrome.
 * 
 * Return all possible palindrome partitioning of s.
 * 
 * For example, given s = "aab",
 * Return
 * 
 *   [
 *     ["aa","b"],
 *     ["a","a","b"]
 *   ]
 */

import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<List<String>>();
        List<String> path = new ArrayList<String>();
        partitionHelper(s, 0, path, result);
        return result;
    }

    private void partitionHelper(String s, int start, List<String> path, List<List<String>> result) {
        if (start == s.length()) {
            result.add(new ArrayList<String>(path));
            return;
        }

        for (int end = start + 1; end < s.length(); ++end) {
            if (isPalindrome(s, start, end - 1)) {
                path.add(s.substring(start, end));
                partitionHelper(s, end, path, result);
                path.remove(path.size() - 1);
            }
        }
    }

    private isPalindrome(String s, int start, int end) {
        while (start < end) {
            if (s.charAt(start) != s.charAt(end)) {
                return false;
            }
            ++start;
            --end;
        }

        return true;
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/palindrome-partitioning.ts
/*
Given a string s, partition s such that every substring of the partition is a palindrome. Return all possible palindrome partitioning of s.

A palindrome string is a string that reads the same backward as forward.



Example 1:

Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]
Example 2:

Input: s = "a"
Output: [["a"]]


Constraints:

1 <= s.length <= 16
s contains only lowercase English letters.
*/

function palindromePartition(input: string): string[][] {
  const result: string[][] = [];
  const path: string[] = [];
  palindromePartitionDfs(0);
  return result;

  function palindromePartitionDfs(start: number): void {
    if (start === input.length) {
      result.push([...path]);
    }

    for (let i = start; i < input.length; ++i) {
      if (isPalindrome(start, i)) {
        path.push(input.substring(start, i + 1));
        palindromePartitionDfs(i + 1);
        path.pop();
      }
    }
  }

  function isPalindrome(left: number, right: number): boolean {
    for (; left < right; ++left, --right) {
      if (input[left] !== input[right]) {
        return false;
      }
    }
    return true;
  }
}
